File: frontend/src/lib/hackathon.ts
// Typed client for the hackathon module API (S7.3).
// Self-contained request helper with an injectable fetch so every path is unit-testable.

export type SubmissionStatus = "draft" | "submitted" | "disqualified";

export interface TeamMember {
  user_id: string;
  email: string;
  full_name: string;
}

export interface Team {
  id: string;
  event_id: string;
  name: string;
  members: TeamMember[];
}

export interface Submission {
  id: string;
  team_id: string;
  title: string;
  summary: string;
  repo_url: string;
  demo_url: string;
  status: SubmissionStatus;
}

export interface SubmissionInput {
  title: string;
  summary?: string;
  repo_url?: string;
  demo_url?: string;
}

export interface LeaderboardRow {
  team_id: string;
  team_name: string;
  submission_id: string;
  submission_title: string;
  status: SubmissionStatus;
  total_score: number;
  average_score: number;
  score_count: number;
  rank: number;
}

const API_BASE = import.meta.env.VITE_API_BASE ?? "";

async function request<T>(
  path: string,
  init: RequestInit,
  fetchImpl: typeof fetch,
): Promise<T> {
  const resp = await fetchImpl(`${API_BASE}${path}`, {
    headers: { "Content-Type": "application/json" },
    ...init,
  });
  if (!resp.ok) {
    let detail = `Request failed: ${resp.status}`;
    try {
      const body = await resp.json();
      if (body && typeof body.detail === "string") detail = body.detail;
    } catch {
      // non-JSON error body — keep the status-based message
    }
    throw new Error(detail);
  }
  return (await resp.json()) as T;
}

function base(eventId: string): string {
  return `/api/events/${eventId}/hackathon`;
}

export function listTeams(
  eventId: string,
  fetchImpl: typeof fetch = fetch,
): Promise<Team[]> {
  return request<Team[]>(`${base(eventId)}/teams`, { method: "GET" }, fetchImpl);
}

export function createTeam(
  eventId: string,
  name: string,
  fetchImpl: typeof fetch = fetch,
): Promise<Team> {
  return request<Team>(
    `${base(eventId)}/teams`,
    { method: "POST", body: JSON.stringify({ name }) },
    fetchImpl,
  );
}

export function joinTeam(
  eventId: string,
  teamId: string,
  fetchImpl: typeof fetch = fetch,
): Promise<Team> {
  return request<Team>(
    `${base(eventId)}/teams/${teamId}/join`,
    { method: "POST" },
    fetchImpl,
  );
}

export function listSubmissions(
  eventId: string,
  fetchImpl: typeof fetch = fetch,
): Promise<Submission[]> {
  return request<Submission[]>(
    `${base(eventId)}/submissions`,
    { method: "GET" },
    fetchImpl,
  );
}

export function createSubmission(
  eventId: string,
  teamId: string,
  input: SubmissionInput,
  fetchImpl: typeof fetch = fetch,
): Promise<Submission> {
  return request<Submission>(
    `${base(eventId)}/teams/${teamId}/submission`,
    { method: "POST", body: JSON.stringify(input) },
    fetchImpl,
  );
}

export function updateSubmission(
  eventId: string,
  submissionId: string,
  input: Partial<SubmissionInput> & { status?: SubmissionStatus },
  fetchImpl: typeof fetch = fetch,
): Promise<Submission> {
  return request<Submission>(
    `${base(eventId)}/submissions/${submissionId}`,
    { method: "PATCH", body: JSON.stringify(input) },
    fetchImpl,
  );
}

export function getLeaderboard(
  eventId: string,
  fetchImpl: typeof fetch = fetch,
): Promise<LeaderboardRow[]> {
  return request<LeaderboardRow[]>(
    `${base(eventId)}/leaderboard`,
    { method: "GET" },
    fetchImpl,
  );
}

File: backend/app/schemas/hackathon.py
"""Request/response schemas for the hackathon module (S7.3)."""
from __future__ import annotations

from pydantic import BaseModel, ConfigDict, Field

from app.models.hackathon import SubmissionStatus


# ---------- teams ----------

class TeamCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)


class TeamMemberRead(BaseModel):
    user_id: str
    email: str
    full_name: str


class TeamRead(BaseModel):
    id: str
    event_id: str
    name: str
    members: list[TeamMemberRead] = []


# ---------- submissions ----------

class SubmissionCreate(BaseModel):
    title: str = Field(min_length=1, max_length=200)
    summary: str = Field(default="", max_length=5000)
    repo_url: str = Field(default="", max_length=500)
    demo_url: str = Field(default="", max_length=500)


class SubmissionUpdate(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=200)
    summary: str | None = Field(default=None, max_length=5000)
    repo_url: str | None = Field(default=None, max_length=500)
    demo_url: str | None = Field(default=None, max_length=500)
    status: SubmissionStatus | None = None


class SubmissionRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    team_id: str
    title: str
    summary: str
    repo_url: str
    demo_url: str
    status: SubmissionStatus


# ---------- scoring ----------

class ScoreCreate(BaseModel):
    criterion: str = Field(min_length=1, max_length=100)
    value: float = Field(ge=0, le=10)


class ScoreRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    id: str
    submission_id: str
    judge_id: str
    criterion: str
    value: float


# ---------- leaderboard ----------

class LeaderboardRow(BaseModel):
    team_id: str
    team_name: str
    submission_id: str
    submission_title: str
    status: SubmissionStatus
    total_score: float
    average_score: float
    score_count: int
    rank: int
